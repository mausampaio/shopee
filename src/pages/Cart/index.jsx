import React from "react";
import { toast } from "react-toastify";
import { Loading } from "../../components/Loading";
import { Store } from "../../context";
import { Container } from "../../components/Container";
import { Button } from "../../components/Button";
import { priceFormatter } from "../../utils/priceFormatter";
import { CartItem } from "./CartItem";
import { Content, Header, Footer, TotalPrice } from "./styles";
import useWindowDimensions from "../../hooks/useWindowDimensions";
import SearchForm from "../../components/SearchForm";

const Cart = () => {
  const showAlert = () =>
    new Promise((resolve) => {
      const sucess = () => {
        toast.success("A compra foi processada com sucesso, Obrigado!", {
          position: "bottom-center",
        });
        resolve();
      };
      setTimeout(sucess, 2500);
    });

  const context = React.useContext(Store);
  const {
    cart,
    setCart,
    loading,
    setLoading,
    addProduct,
    removeProduct,
    getCartTotal,
    removeCartFromLocalStorage,
  } = context;

  const { width } = useWindowDimensions();

  const checkout = async () => {
    setLoading(true);
    setCart([]);
    removeCartFromLocalStorage();
    await showAlert();
    setLoading(false);
  };

  if (loading) {
    return <Loading />;
  } else {
    return (
      <Container>
        <h1 style={{ margin: "32px 0 36px" }}>Carrinho</h1>
        {width <= 575 ? <SearchForm /> : null}
        {cart.length === 0 ? (
          <h2 style={{ display: "flex", justifyContent: "center" }}>
            Não existem produtos
          </h2>
        ) : (
          <div>
            <Header>
              <div></div>
              <span>Preço UN.</span>
              <span>Quantidade</span>
              <span>Subtotal</span>
            </Header>
            <Content>
              {cart?.map((product, index) => (
                <CartItem
                  key={product.id}
                  product={product}
                  lineBreak={cart.length !== index + 1}
                  onIncrement={addProduct}
                  onDecrement={removeProduct}
                />
              ))}
            </Content>
            <Footer>
              <div>
                <TotalPrice>
                  {priceFormatter({
                    price: getCartTotal(),
                    formatCents: false,
                  })}
                </TotalPrice>
              </div>
              <div style={{ width: "226px" }}>
                <Button onClick={checkout}>Finalizar Compra</Button>
              </div>
            </Footer>
          </div>
        )}
      </Container>
    );
  }
};

export default Cart;
