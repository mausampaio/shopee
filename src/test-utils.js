import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

export const mockProduct = {
  id: 1,
  name: "Produto de teste",
  price: 1.0,
  description: "Produto gerado para testes",
  url: "produto-de-teste",
  image: "./produto-teste.jpg",
};

export const mockCart = {
  id: mockProduct.id,
  name: mockProduct.name,
  price: mockProduct.price,
  quantity: 10,
};

export const mockContext = {
  cart: [],
  products: [],
  product: {},
  loading: false,
  getCartTotal: () => 0,
};

const RouterProvider = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};

const customRender = (ui, options) => {
  return render(<RouterProvider>{ui}</RouterProvider>, { ...options });
};

export * from "@testing-library/react";

export { customRender as render };
