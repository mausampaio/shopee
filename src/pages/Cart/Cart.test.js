import { Store } from "../../context";
import { mockCart, mockContext, render } from "../../test-utils";
import Cart from "./";

const ContextProvider = ({ children }) => (
  <Store.Provider value={mockContext}>{children}</Store.Provider>
);

test('should render "Não existem produtos" when cart is empty', () => {
  const { getByText } = render(<Cart />, { wrapper: ContextProvider });
  expect(getByText("Não existem produtos")).toBeInTheDocument();
});

test("should render products on the cart", () => {
  const MockedCart = () => (
    <Store.Provider value={{ ...mockContext, cart: [mockCart] }}>
      <Cart />
    </Store.Provider>
  );
  const { getByText } = render(<MockedCart />);
  expect(getByText("Produto de teste")).toBeInTheDocument();
});

test("should match snapshot", () => {
  const { asFragment } = render(<Cart />, { wrapper: ContextProvider });
  expect(asFragment()).toMatchSnapshot();
});
