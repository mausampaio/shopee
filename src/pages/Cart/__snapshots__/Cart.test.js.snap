// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 80%;
  max-width: 1280px;
  margin: 0 auto;
}

@media screen and (max-width:768px) {
  .c0 {
    width: 90%;
  }
}

<div
    class="c0"
  >
    <h1
      style="margin: 32px 0px 36px;"
    >
      Carrinho
    </h1>
    <h2
      style="display: flex; justify-content: center;"
    >
      NÃ£o existem produtos
    </h2>
  </div>
</DocumentFragment>
`;
